src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/ImmutableArrayExtensions.cs

```diff
@@ -255,206 +372,315 @@ public static ImmutableArray<T> OrderDescendingAsArray<T>(this ImmutableArray<T>
-        private readonly bool _useComparer;
-        private readonly bool _descending;
+        var compareHelper = new CompareHelper<TKey>(comparer, descending: true);
+        return builder.ToImmutableOrderedByCore(keySelector, in compareHelper);
+    }
 
-        public CompareHelper(IComparer<T>? comparer, bool descending)
-        {
-            _comparerSpecified = comparer is not null;
-            _comparer = comparer ?? Comparer<T>.Default;
-            _useComparer = true;
-            _descending = descending;
-            _comparison = null!;
-        }
+    /// <summary>
+    ///  Returns an immutable array that contains the current contents of this
+    ///  <see cref="ImmutableArray{T}.Builder"/> sorted in descending order according to a key.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the elements in <paramref name="builder"/>.</typeparam>
+    /// <typeparam name="TKey">The type of key returned by <paramref name="keySelector"/>.</typeparam>
+    /// <param name="builder">The <see cref="ImmutableArray{T}.Builder"/> whose contents will be sorted.</param>
+    /// <param name="keySelector">A function to extract a key from an element.</param>
+    /// <param name="comparison">An <see cref="Comparison{T}"/> to compare keys.</param>
+    /// <returns>
+    ///  Returns a new <see cref="ImmutableArray{T}"/> whose elements are sorted in descending order according to a key.
+    /// </returns>
+    public static ImmutableArray<TElement> ToImmutableOrderedByDescending<TElement, TKey>(
+        this ImmutableArray<TElement>.Builder builder, Func<TElement, TKey> keySelector, Comparison<TKey> comparison)
+    {
+        var compareHelper = new CompareHelper<TKey>(comparison, descending: true);
+        return builder.ToImmutableOrderedByCore(keySelector, in compareHelper);
+    }
 
-        public CompareHelper(Comparison<T> comparison, bool descending)
+    private static ImmutableArray<T> ToImmutableOrderedCore<T>(this ImmutableArray<T>.Builder builder, ref readonly CompareHelper<T> compareHelper)
+    {
+        if (builder.Count <= 1)
         {
-            _comparison = comparison;
-            _useComparer = false;
-            _descending = descending;
-            _comparer = null!;
+            return builder.ToImmutable();
         }
 
-        public bool InSortedOrder(T? x, T? y)
+        if (SortHelper.AreOrdered(builder, in compareHelper))
         {
-            // We assume that x and y are in sorted order if x is > y.
-            // We don't consider x == y to be sorted because the actual sor
-            // might not be stable, depending on T.
-
-            return _useComparer
-                ? !_descending ? _comparer.Compare(x!, y!) > 0 : _comparer.Compare(y!, x!) > 0
-                : !_descending ? _comparison(x!, y!) > 0 : _comparison(y!, x!) > 0;
+            // No need to sort - items are already ordered.
+            return builder.ToImmutable();
```

Comment:
Can we call `DrainToImmutable` here?

Reply:
I thought about that and decided that it might be surprising to use `DrainToImmutable` within `ToImmutable` methods. I was thinking of adding `DrainToImmutableOrdered*` variants. What do you think?

Reply:
Sounds good to me.

Reply:
Done